lua/iswap.lua

```diff
@@ -262,58 +262,51 @@ function M.iswap_node(config, direction)
   vim.cmd([[silent! call repeat#set("\<Plug>ISwapNormal", -1)]])
 end
 
-
 -- TODO: refactor iswap() and iswap_with()
 -- swap current with one other node
-function M.iswap_with(direction, config)
+function M.imove_with(direction, config) M.iswap_with(direction, config, true) end
+function M.imove(direction, config) M.iswap(config, true) end
+
+function M.iswap_with(direction, config, imove)
   config = M.evaluate_config(config)
   local bufnr = vim.api.nvim_get_current_buf()
   local winid = vim.api.nvim_get_current_win()
 
-  local parent = internal.get_list_node_at_cursor(winid, config)
-  if not parent then
+  local parent, children, cur_node_idx = internal.get_list_node_at_cursor(winid, config, true)
+  if not parent or not children or not cur_node_idx then
     err('did not find a satisfiable parent node', config.debug)
     return
   end
-  local children = ts_utils.get_named_children(parent)
-
-  -- nothing to swap here
-  if #children < 2 then return end
-
-  local cur_nodes = util.nodes_containing_cursor(children, winid)
-  if #cur_nodes == 0 then
-    err('not on a node!', 1)
-  end
-
-  if #cur_nodes > 1 then
-    err('multiple found, using first', config.debug)
-  end
 
-  local cur_node = children[cur_nodes[1]]
-  table.remove(children, cur_nodes[1])
+  local cur_node = table.remove(children, cur_node_idx)
 
   local sr, sc, er, ec = parent:range()
 
-  -- a is the node to swap the cur_node with
-  local a
+  -- a is the node to move the cur_node into the place of
+  local a, a_idx
 
   -- enable autoswapping with one other child
   -- and default to prompting for user input
   if config.autoswap and #children == 1 then
     a = children[1]
   else
     if direction == 'left' then
-      a = children[cur_nodes[1] - 1]
+      a = children[cur_node_idx - 1]
     elseif direction == 'right' then
       -- already shifted over, no need for +1
-      a = children[cur_nodes[1]]
+      a = children[cur_node_idx]
     else
-      local user_input = ui.prompt(bufnr, config, children, {{sr, sc}, {er, ec}}, 1)
+      local user_input, user_input_idx = ui.prompt(bufnr, config, children, { { sr, sc }, { er, ec } }, 1)
```

Comment:
Would it be cleaner if `ui.prompt` returned *only* the indices? Since the node(s) are easily computable from the indices.

Reply:
That does sound cleaner. I was in a bit of a "maintain backwards compatibility so I have to touch as little of the rest of the codebase" mode to just get it working